File: Python/15reverse_list.py



# iterative
# n = number of nodes
# Time: O(n)
# Space: O(1)

def reverse_list(head):
  prev = None
  current = head
  while current is not None:
    next = current.next
    current.next = prev
    prev = current
    current = next
  return prev

# recursive
# n = number of nodes
# Time: O(n)
# Space: O(n)

def reverse_list(head, prev = None):
  if head is None:
    return prev
  next = head.next
  head.next = prev
  return reverse_list(next, head)

File: Javascript/15reverseList.js

// RECURSIVE
// n = # of nodes
// time:  O(n);
// space: O(1);

const reverseList = (head) => {
  // todo
  let prev = null;
  let current = head;
  
  while (current) {
    
    let temp = current.next;
    current.next = prev;
    prev = current;
    current = temp;
  }
  
  return prev;
};


// RECURSIVE
// n = # of nodes
// time:  O(n);
// space: O(n);
const reverseList = (head, prev=null) => {
  // todo
  if (!head) return prev;
  let nextHead = head.next;
  head.next = prev;
  return reverseList(nextHead, head);
};